src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -0,0 +1,325 @@
+                else:
+                    raise
+            ctx.iam.delete_instance_profile(profile_name)
+
+    def _addNodes(self, instancesToLaunch, preemptable=None):
+        bdm = self._getBlockDeviceMapping()
+        profileID = str(time.time())
+        arn = self._getProfileARN(self.ctx, instanceID=profileID, role='worker')
+        workerData = {'role': 'worker', 'tag': workerTag, 'args': workerArgs.format(self.masterIP)}
+        userData = AWSUserData.format(**workerData)
+        if not preemptable:
+            logger.debug('Launching non-preemptable instance(s)')
+            for attempt in retry_ec2(retry_while=expectedLaunchErrors):
+                with attempt:
+                    self.ctx.ec2.run_instances(image_id=coreOSAMI, min_count=instancesToLaunch,
+                                               max_count=instancesToLaunch, key_name=self.keyName,
+                                               security_groups=[self.securityGroupName],
+                                               instance_type=self.instanceType.name,
+                                               instance_profile_arn=arn, user_data=userData,
+                                               block_device_map=bdm)
+        else:
+            logger.debug('Launching spot instance(s) with bid of %s', self.spotBid)
+            requests = []
+            for attempt in retry_ec2(retry_while=expectedLaunchErrors):
+                with attempt:
+                    # returns list of SpotInstanceRequests
+                    requests = self.ctx.ec2.request_spot_instances(price=self.spotBid, image_id=coreOSAMI,
+                                                                   count=instancesToLaunch, key_name=self.keyName,
+                                                                   security_groups=[self.securityGroupName],
+                                                                   instance_type=self.instanceType.name,
+                                                                   instance_profile_arn=arn, user_data=userData,
+                                                                   block_device_map=bdm)
+            self._waitOnRequests(self.ctx, requests)
+
+        logger.info('Launched %s new instance(s)', instancesToLaunch)
+
+    def _getBlockDeviceMapping(self):
+        # determine number of ephemeral drives via cgcloud-lib
+        bdtKeys = ['', '/dev/xvdb', '/dev/xvdc', '/dev/xvdd']
+        bdm = BlockDeviceMapping()
+        # the first disk is already attached for us so start with 2nd.
+        for disk in xrange(1, self.instanceType.disks + 1):
+            bdm[bdtKeys[disk]] = BlockDeviceType(
+                ephemeral_name='ephemeral{}'.format(disk - 1))  # ephemeral counts start at 0
+
+        logger.debug('Device mapping: %s', bdm)
+        return bdm
+
+    @staticmethod
+    def _waitOnRequests(ctx, requests):
```

Comment:
This might be too simplistic. Why don't we use cgcloud-lib's wait_spot_requests_active? They have timeouts and the tentative logic that helped us a great deal during the TCGA recompute. I don't think we want to return to the scaler until the instances are running, so wait_instances_running() should also be used.